### PR TITLE
Bypass LAN and reserved subnets in the VPN

### PR DESCRIPTION
**Pull request by alalamav**
* Excludes [special registry subnets](https://iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml) from the VPN. This includes LAN and reserved subnets.
  * Apple: NetworkExtension APIs support excluding subnets from the VPN.
  * Android: VpnService APIs support only subnet inclusion, so we have computed the subnet compliment of the reserved subnets.
  * Windows: OutlineService adds direct routes to the default gateway for the bypassed subnets.

### CODE REVIEW

File: tools/OutlineService/OutlineService/OutlineService.cs [hunk 1]
```diff
@@ -383,6 +406,29 @@ public partial class OutlineService : ServiceBase {
       }
     }
 
+    // Routes reserved and private subnets through the default gateway so they bypass the VPN.
+    private void AddReservedSubnetBypass(string systemGatewayIp, string interfaceName) {
+      try {
+        foreach (string subnet in IPV4_RESERVED_SUBNETS) {
```
Comment (trevj) [inline tools/OutlineService/OutlineService/OutlineService.cs]: Let's think about exception handling. To date, `ConfigureRouting` has thrown an exception if **any** step fails - that's not the case here, as you're swallowing the exception.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java [hunk 2]
```diff
@@ -70,15 +73,22 @@ public boolean establishVpn() {
...
+              .addAddress(String.format(Locale.ROOT, VPN_INTERFACE_PRIVATE_LAN, "1"),
+                  VPN_INTERFACE_PREFIX_LENGTH)
+              .addDnsServer(dnsResolverAddress)
+              .addDisallowedApplication(vpnService.getPackageName());
+
+      // In absence of an API to remove routes, instead of adding the default route (0.0.0.0/0),
+      // retrieve the list of subnets that excludes those reserved for special use.
+      final ArrayList<Subnet> reservedBypassSubnets = getReservedBypassSubnets();
+      for (Subnet subnet : reservedBypassSubnets) {
+        builder.addRoute(subnet.address, subnet.prefix);
```
[the remark this answers is not in the document]
Comment (alalamav) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java]: It's not necessary to get the builder reference - works as is.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java [hunk 3]
```diff
@@ -157,4 +167,42 @@ public void disconnectTunnel() {
   private String selectDnsResolverAddress() {
     return DNS_RESOLVER_IP_ADDRESSES[new Random().nextInt(DNS_RESOLVER_IP_ADDRESSES.length)];
   }
+
+  /* Returns a subnet list that excludes reserved subnets. */
+  private ArrayList<Subnet> getReservedBypassSubnets() {
+    final String[] subnetStrings = vpnService.getResources().getStringArray(
+        vpnService.getResourceId(PRIVATE_LAN_BYPASS_SUBNETS_ID, "array"));
```
[the remark this answers is not in the document]
Comment (alalamav) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java]: Unfortunately, we cannot access the generated R class from Cordova plugins. Therefore we resort to retrieving resources via the Resources class.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java [hunk 4]
```diff
@@ -530,7 +530,7 @@ private void startForegroundWithNotification(
       builder = new Notification.Builder(this);
     }
     try {
-      builder.setSmallIcon(getResourceIdForDrawable("small_icon"));
+      builder.setSmallIcon(getResourceId("small_icon", "drawable"));
```
[the remark this answers is not in the document]
Comment (alalamav) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java]: See answer above.

File: tools/OutlineService/OutlineService/OutlineService.cs (hunk 1 above)
Comment (alalamav) [reply tools/OutlineService/OutlineService/OutlineService.cs]: This was intentional - I don't think that it's worth failing a connection if we fail to exclude a particular subnet. I'm not sure that there is an equivalent step in `ConfigureRouting` at this point; if I'm not mistaken they are all necessary for the VPN to operate correctly, with setting the lower interface metric on the fence.

What would be the benefit of throwing and failing the connection?

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java [hunk 5]
```diff
@@ -157,4 +167,42 @@ public void disconnectTunnel() {
   private String selectDnsResolverAddress() {
     return DNS_RESOLVER_IP_ADDRESSES[new Random().nextInt(DNS_RESOLVER_IP_ADDRESSES.length)];
   }
+
+  /* Returns a subnet list that excludes reserved subnets. */
```
[the remark this answers is not in the document]
Comment (alalamav) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java]: I have been using `/**/` for method and class level comments, and `//` for comments in the code. I can change them all to long comments.

File: tools/OutlineService/OutlineService/OutlineService.cs (hunk 1 above)
Comment (trevj) [reply tools/OutlineService/OutlineService/OutlineService.cs]: The benefit is that we don't leave the client in a weird state. I'm okay with that for this but let's update the comments for `ConfigureRouting`.

File: cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bemasc) [reply cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnel.java]: In that case, I would recommend using `/** */` in Javadoc style.

**Pull request merged**